### ISSUE TITLE
How to allow special characters in my input query params

### ISSUE THREAD

**Issue by SriramZafin** [swagger-api/swagger-ui#3915]
I have a rest end point for which a valid URL will be like the below -

/app/api/rest?formatType=json&opName=servicerequest:userAPIGetChildServiceRequests&opData={param0:{"requestId":1000}}

I setup the "opData" & "opName" as query IN parameters, so that I will get a text box with a default value like the below -

opName - servicerequest:userAPIGetChildServiceRequests
opData - {param0:{'requestId':1000}}

The problem is when I hit the execute button, the special characters like {,[,: get encoded and the resulting encoding URL

/app/api/rest?formatType=json&opName=servicerequest%3AuserAPIGetChildServiceRequests&opData=%7Bparam0%3A%7B'requestId'%3A1000%7D%7D

 is invalid when it reaches my rest endpoint.

I need a way to disable this encoding so that the URL when it reaches my end point would be the same as

/app/api/rest?formatType=json&opName=servicerequest:userAPIGetChildServiceRequests&opData={param0:{"requestId":1000}}

I am using Swagger UI 3.4.4, I can even try out specific versions of Swagger UI if that will solve my problem.

Any pointers to get past this issue or any workarounds will be greatly helpful
Let me know if you need more information

| Q                               | A
| ------------------------------- | -------
| Bug or feature request?         |  
| Which Swagger/OpenAPI version?  |
| Which Swagger-UI version?       |  3.4.4
| How did you install Swagger-UI? |  From GITHUB
| Which browser & version?        | Chrome version 62
| Which operating system?         | Windows

**Comment by shockey** [comment]
Hi @SriramZafin - are you writing a Swagger 2.0 or OpenAPI 3.0 API definition?

**Comment by SriramZafin** [comment]
I need to provide Swagger specs for the REST APIs we support.

I started with writing Swagger 2.0 definitions, then took the latest of Swagger UI(Swagger 3.4.4) and started defining for that.

**Comment by webron** [comment]
@SriramZafin we can provide partial support to what you need, but not all of it. Here's the breakdown.

Not encoding characters like `:` is doable only with OAS3, using the [`allowReserved`](https://github.com/OAI/OpenAPI-Specification/blob/OpenAPI.next/versions/3.0.0.md#parameterAllowReserved) keyword. There's no way to control it in Swagger 2.0.

Not encoding curly braces (`{ }`) will never be supported as those characters are not allowed in URLs. Here are two references:
 - https://tools.ietf.org/html/rfc1738#section-2.2 under `Unsafe`:
```
   Characters can be unsafe for a number of reasons.  The space
   character is unsafe because significant spaces may disappear and
   insignificant spaces may be introduced when URLs are transcribed or
   typeset or subjected to the treatment of word-processing programs.
   The characters "<" and ">" are unsafe because they are used as the
   delimiters around URLs in free text; the quote mark (""") is used to
   delimit URLs in some systems.  The character "#" is unsafe and should
   always be encoded because it is used in World Wide Web and in other
   systems to delimit a URL from a fragment/anchor identifier that might
   follow it.  The character "%" is unsafe because it is used for
   encodings of other characters.  Other characters are unsafe because
   gateways and other transport agents are known to sometimes modify
   such characters. These characters are "{", "}", "|", "\", "^", "~",
   "[", "]", and "`".

   All unsafe characters must always be encoded within a URL. For
   example, the character "#" must be encoded within URLs even in
   systems that do not normally deal with fragment or anchor
   identifiers, so that if the URL is copied into another system that
   does use them, it will not be necessary to change the URL encoding.
```

- https://tools.ietf.org/html/rfc3986#section-2.2 

Since those characters are not allowed in URLs, we have to percent-encode them no matter what.

**Comment by sithmein** [comment]
Interestingly curly braces are *not* escaped with version 3.4.3 and OAS3. The request URL will look like

`http://localhost:8080/com.knime.enterprise.server/rest/v4/repository/OpenAPI%20Test?json-input={%22a%22:%22b%22}`

which will be rejected by e.g. Tomcat.

**Comment by webron** [comment]
@sithmein interesting, that's something we'd need to fix. @shockey?

**Comment by shockey** [comment]
Interesting indeed - I fixed an `allowReserved` issue last week (via https://github.com/swagger-api/swagger-js/pull/1181) and started using a library that encodes according to RFC3986... but we aren't encoding characters that are neither 3986 Reserved or Unreserved (for example, `{` and `}`) regardless of `allowReserved`'s value.

Issue opened here: https://github.com/swagger-api/swagger-js/issues/1184

**Comment by teaearlgraycold** [comment]
Not sure if this is the same issue, but I'm unable to send `&` in my query parameters. Shouldn't the strings all be percent encoded before being sent out?

This is with OpenAPI 3. Not sure how to check the swagger UI version but it's probably 3.4.3.

**Comment by shockey** [comment]
@teaearlgraycold 

> I'm unable to send & in my query parameters

What's happening? I'm reading this as "I expect `&` to be escaped but it's being sent as-is", please let me know if I'm on the right track.

> Shouldn't the strings all be percent encoded

Yep, this is covered in https://github.com/swagger-api/swagger-js/issues/1184 - basically, right now we're handling the "allowable" reserved characters in RFC3986 correctly, but other special characters (including `&`) are not being escaped correctly.

**Comment by teaearlgraycold** [comment]
>I'm reading this as "I expect & to be escaped but it's being sent as-is", please let me know if I'm on the right track.

Yes, you're reading that correctly.

**Comment by shockey** [comment]
Closing this: it's been inactive for a week, questions above have been addressed, and the bug mentioned is tracked in #1184 (should be fixed this week). 

If there's anything else, feel free to bump this thread and I'll reopen!

**Comment by sveillette** [comment]
I was simply trying to enter an email address in a text field:

I wanted testuser@xyz.com
The value presented to my service was testuser@xyz

So to get the whole email address i entered the email address with a **trailing** dot character

testuser@xyz.com.

swagger will present this value as: testuser@xyz.com
